### PR TITLE
Write Patch, Read and Replace DeploymentStatus test +1

### LINKED ISSUE #102258: Write Patch, Read and Repleace DeploymentStatus test +1
# Progress <code>[4/6]</code>

-   [X] APISnoop org-flow: [Appsv1DeploymentStatusLifecycleTest.org](https://github.com/apisnoop/ticket-writing/blob/master/Appsv1DeploymentStatusLifecycleTest.org)
-   [X] Test approval issue: [kubernetes/kubernetes#102258](https://github.com/kubernetes/kubernetes/issues/102258)
-   [X] Test PR: kuberenetes/kubernetes #102260
-   [X] Two weeks soak start date: 17 Jun 2021 [Testgrid](https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should.validate.Deployment.Status.endpoints&graph-metrics=test-duration-minutes)
-   [ ] Two weeks soak end date: 1 July 2021
-   [ ] Test promotion PR: kubernetes/kubernetes#?

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining StatefulSet endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%DeploymentStatus'
      order by kind, endpoint desc
      limit 10;
```

```example
                  endpoint                 |                              path                              |    kind
  -----------------------------------------+----------------------------------------------------------------+------------
   replaceAppsV1NamespacedDeploymentStatus | /apis/apps/v1/namespaces/{namespace}/deployments/{name}/status | Deployment
  (1 row)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API / Workload Resources / Deployment](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/deployment-v1/)
-   [client-go - Deployment](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/apps/v1/deployment.go)

# The mock test

## Test outline

1.  Create a watch to track deployment events.

2.  Create a deployment with a static label. Confirm that the pods are running.

3.  Get the deployment status. Parse the response and confirm that the deployment status conditions can be listed.

4.  Update the deployment status. Confirm via the watch that the status has been updated.

5.  Patch the deployment status. Confirm via the watch that the status has been patched.

## Test the functionality in Go

Using an existing [status lifecycle test](https://github.com/ii/kubernetes/blob/ca3aa6f5af1b545b116b52c717b866e43c79079b/test/e2e/apps/daemon_set.go#L812-L947) as a template for a new [ginkgo test](https://github.com/ii/kubernetes/blob/deployment-status-test/test/e2e/apps/deployment.go#L487-L621) for deployment lifecycle test.

# Verifying increase in coverage with APISnoop

## Listing endpoints hit by the new e2e test

This query shows the endpoints hit within a short period of running the e2e test

```sql-mode
select distinct  endpoint, right(useragent,65) AS useragent
from testing.audit_event
where endpoint ilike '%DeploymentStatus%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%should%'
order by endpoint
limit 10;
```

```example
                endpoint                 |                             useragent
-----------------------------------------+-------------------------------------------------------------------
 patchAppsV1NamespacedDeploymentStatus   | [sig-apps] Deployment should validate Deployment Status endpoints
 readAppsV1NamespacedDeploymentStatus    | [sig-apps] Deployment should validate Deployment Status endpoints
 replaceAppsV1NamespacedDeploymentStatus | [sig-apps] Deployment should validate Deployment Status endpoints
(3 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 1 point**

This test is also created with the goal of conformance promotion.

---
/sig apps
/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceAppsV1NamespacedDeploymentStatus

**Which issue(s) this PR fixes:**
Fixes #102258

**Testgrid Link:**
[Testgrid](https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should.validate.Deployment.Status.endpoints&graph-metrics=test-duration-minutes)

**Special notes for your reviewer:**
Adds +1 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
```

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd

Unreleated flake
```
 [Fail] [sig-instrumentation] MetricsGrabber [It] should grab all metrics from a Scheduler. 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:88
```

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
Unrelated flake
```
I0524 19:04:57.305] • Failure [65.512 seconds]
I0524 19:04:57.305] [sig-instrumentation] MetricsGrabber
I0524 19:04:57.305] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/common/framework.go:23
I0524 19:04:57.306]   should grab all metrics from a Scheduler. [It]
I0524 19:04:57.306]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:81
I0524 19:04:57.306] 
I0524 19:04:57.306]   May 24 19:04:52.365: Unexpected error:
```

**Comment by oomichi** [comment]
Nice additional test.

/triage accepted
/lgtm

**Comment by Riaankl** [comment]
/assign @soltysh @smarterclayton
This test follows much the same pattern used in the other Apps e2e & Conformance test to clean up the last technical debt.

File: test/e2e/apps/deployment.go [hunk 1]
```diff
@@ -481,6 +484,140 @@ var _ = SIGDescribe("Deployment", func() {
...
+		defer cancel()
+
+		_, err = watchtools.Until(ctx, dList.ResourceVersion, w, func(event watch.Event) (bool, error) {
+			if d, ok := event.Object.(*appsv1.Deployment); ok {
+				found := d.ObjectMeta.Name == testDeployment.ObjectMeta.Name &&
+					d.ObjectMeta.Namespace == testDeployment.ObjectMeta.Namespace &&
+					d.Labels["e2e"] == "testing"
+
+				if !found {
+					framework.Logf("Observed Deployment %v in namespace %v with annotations: %v & Conditions: %v\n", d.ObjectMeta.Name, d.ObjectMeta.Namespace, d.Annotations, d.Status.Conditions)
```
Comment (ravisantoshgudimetla) [inline test/e2e/apps/deployment.go:564]: nit: Same as the comment I mentioned at https://github.com/kubernetes/kubernetes/pull/102256#discussion_r646206878

**Review by ravisantoshgudimetla — APPROVED** [review]
LGTM expect a minor nit.

/lgtm
/hold

Please feel free to remove the hold if you don't want to address the nit

File: test/e2e/apps/deployment.go (hunk 1 above)
Comment (heyste) [reply test/e2e/apps/deployment.go:564]: @ravisantoshgudimetla If the test was to flake at some point the idea was to have a clear reference point to help debug the test. For some context refer to this [comment](https://github.com/kubernetes/kubernetes/pull/96485#issuecomment-732335344).

Comment (Riaankl) [reply test/e2e/apps/deployment.go:564]: The same pattern is followed in other merged conformance tests. We would like to keep it consistent. Thank you for the review.

**Comment by Riaankl** [comment]
/unhold
Thank you @ravisantoshgudimetla.
The same pattern is followed in other merged conformance tests. We would like to keep it consistent. Thank you for the review.

**Review by soltysh — APPROVED** [review]
/priority backlog
/approve

**Pull request merged**